### ISSUE TITLE
Support Azure CLI authentication in Fluent library 

### ISSUE THREAD

**Issue by isaacabraham** [Azure/azure-libraries-for-net#109]
If using the Azure SDK within the context of e.g. VSTS - which has support for Azure authentication built-in for e.g. Powershell and Azure CLI - it would be very useful to be able to take the "transient" authentication from there. This is important for us as we want to be able to launch a .NET program during a release which will, among other things, perform an ARM deployment.

At the moment, the only way to get a handle to `IAzure` seems to be to do an explicit authentication in code.

**Comment by hovsepm** [comment]
you can achieve this by providing you own Credentials class starting v.1.10. e.g.
```
var client = RestClient
                    .Configure()
                    .WithEnvironment(AzureEnvironment.AzureGlobalCloud)
                    .WithLogLevel(HttpLoggingDelegatingHandler.Level.Basic)
                    .WithCredentials(new AzureCredentials(new TokenCredentials(armAuthToken), 
new AzureCredentials(new TokenCredentials(graphAuthToken), tenantId, AzureEnvironment.AzureGlobalCloud))
                    .Build();
var authenticatedClient = Azure.Authenticate(client, tenantId);
```

**Comment by hovsepm** [comment]
@isaacabraham please open a new issue if the proposed solution does not work for your case.

**Comment by isaacabraham** [comment]
@hovsepm Thanks for getting back here with some guidance - it's much appreciated. However, I would rather you didn't close this issue just yet. Could you first provide some more details on this code snippet, such as explanation of what `tenantId`, `armAuthToken`, `graphAuthToken`, `customTokenProvider` are, where they come from and how should they be stored etc.?

Also, as an aside, I'm not sure why you've closed the issue in the first place without some common consensus here (I think that's normally just good practice). Also, why do you want me to open a *new* issue if the suggestion doesn't work, even though it would be for the exact same issue?

Cheers

**Comment by forki** [comment]
@isaacabraham my theory: it's spring cleaning and people get paid for number of closed issues.

**Comment by hovsepm** [comment]
@forki neah :) We got paid by number of up-votes and likes...
@isaacabraham Fluent SDK provides a token acquisition, caching and renewal out of the box. But in many cases (like the one you have described above) all these procedures are done outside Fluent calls. This new functionality enables those scenarios. In other words you can get a Bearer token any possible way and provide a raw token to Fluent SDK for the further communication with Azure. In that case we will not do any auth call anymore. If you need to access Graph functionality (e.g. Role Assignments etc) you will also need to get Bearer token from the Graph endpoint of azure. 
TenantId should match the one that was used to get Bearer tokens. e.g. here is how you can [see it in the portal](https://stackoverflow.com/questions/26384034/how-to-get-the-azure-account-tenant-id).

**Comment by isaacabraham** [comment]
@hovsepm Thanks for getting back with more info! So are you saying that I can create an `AzureCredentials` object with a TenantId and not have to pass in any TokenCredentials? Or, where would they come from?

**Comment by hovsepm** [comment]
@isaacabraham TenantId should be provided along with token. You can take a look to the TokenProvider class here https://github.com/Azure/azure-sdk-for-net/blob/psSdkJson6/src/SdkCommon/ClientRuntime/ClientRuntime/TokenCredentials.cs#L15 

Simply if you an obtain the token string from the Authentication endpoint (ARM or Grpah) then you can provide it directly to TokenCredentials constructor. In the example above `armAuthToken` and `graphAuthToken` are raw string type tokens acquired by client application. Currently we do not have a convenience layer in Fluent SDK that will let you use Azure CLI or PS for authentication.

**Comment by hovsepm** [comment]
@isaacabraham I've change the title of the issue to better align with your ask. LMK if you think it is still not correct.

**Comment by Alexei-B** [comment]
I feel like the workaround described in this issue still isn't complete. How does one go about obtaining the token(s) from a connected account from PowerShell while in a .NET library? Looking in my user folder there are suspect files (AzurePSDataCollectionProfile.json & TokenCache.dat) but these don't look trivial to process.

**Comment by yaohaizh** [comment]
@xccc-msft, we have already support the CLI authentication. 
- Please provide example for how to using CLI authentication in our sample project. 
- Investigate whether we support the Powershell authentication.

**Comment by xccc-msft** [comment]
We have detailed info for different approaches to authenticate through Azure CLI, which can be found via [AUTH.md](https://github.com/Azure/azure-libraries-for-net/blob/master/AUTH.md). Here is an example how to use file-based authentication with service principal management,  via [AUTH sample](https://github.com/Azure/azure-libraries-for-net/blob/master/Samples/GraphRbac/ManageServicePrincipal.cs).

For Powershell side, we don't have similar way to do such authentication.

For future plan, we will have single sign on to authenticate instead of above approaches.

**Comment by xccc-msft** [comment]
Close this as released this feature in Fluent SDK 1.28.0
